src/rt/sched/backpressure.h

```diff
@@ -0,0 +1,286 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+/**
+ * ## Introduction
+ *
+ * The primary goal of backpressure is to prevent runaway growth of a cown's
+ * message queue that might eventually result in a Verona application running
+ * out of memory.
+ *
+ * The following expects the reader to be familiar with scheduler threads,
+ * work stealing, fairness, cown reference counting, multi-message acquisition,
+ * and actions.
+ *
+ * If a cown receives messages more quickly than it is able to process them, the
+ * queue will start to grow in an unbounded fashion. Cowns that aren't able to
+ * keep up with their arriving messages are marked as "overloaded". Once a cown
+ * is marked as overloaded, cowns sending messages to it will be temporarily
+ * descheduled ("muted"). Descheduling cowns that are sending to an overloaded
+ * cown helps alleviate pressure on the it by:
+ *   - Temporarily removing a source of incoming messges
+ *   - Allowing more scheduler time to be spent on the overload cown rather than
+ *     the descheduled ones.
+ *
+ * Similarly, cowns sending to cowns that have been muted will also be muted to
+ * prevent runaway queue growth while the muted cown in descheduled.
+ *
+ * Muted cowns will be rescheduled after the receiver that triggered their
+ * muting (the "mutor") is no longer muted or overloaded.
+ *
+ * ## Detecting Overloaded Cowns
+ *
+ * An overloaded cown is one that experiences message queue growth as it is
+ * processing messages. Since it would be prohibitively expensive to calculate
+ * the length of a message queue directly, a count of messages removed from the
+ * queue over some period is used as an approximation for message queue size.
+ * Like scheduler fairness, the period is determined using a token message that
+ * will occasionally fall out of the queue and then be inserted to begin the
+ * next period. The "load" of a cown at any time is the accumulation of messages
+ * processed over multiple periods.
+ *
+ * ## Muting Cowns
+ *
+ * All messages from one set of sending cowns to a set of receiving cowns are
+ * scanned to determine if the senders should be muted. The senders will be
+ * muted once their action completes if all of the following are true:
```

Comment:
This is for the avoidance of deadlock.  But you don't mention deadlock at all in the description.  I think it would be good to put an example of how this breaks deadlocks here.  My understanding is:

> * Deadlock could occur if a cown 'A' is muted due to another cown 'B' being overloaded.
> * To process `B`'s messages requires `A`.
>
> Thus, B can no longer proceed.
>
>  In this algorithm, if you are attempting a acquire a muted cown, if you are overloaded, then you `unmute` the cown you are trying to acquire.

Now, there are bunch of awkward corner cases,
* `unmute` involves a message, can a cown be unmuted and remuted, before the `when` is processed.
* can multiple `unmute` messages to the same cown exist.
* What happens if a cown is unmuted, before the message arrives?
* What stops a cown being muted after a multimessage has been sent, that could cause a deadlock?